[JSC] `Uint8Array#setFromBase64` should decode and write chunks which occur prior to bad data

sosukesuzuki opened the pull request:
#### 77046030809c04655dc91b43a43715b43235033e
<pre>
[JSC] `Uint8Array#setFromBase64` should decode and write chunks which occur prior to bad data
<a href="https://bugs.webkit.org/show_bug.cgi?id=276859">https://bugs.webkit.org/show_bug.cgi?id=276859</a>

Reviewed by Yusuke Suzuki.

According to the spec[1][2] and the test[3], Uint8Array.prototype.setFromBase64 throws a SyntaxError
when encountering bad data but writes the decoded chunks prior to bad data.

However, the current JSC discards the decoded data upon encountering bad data and throws an error.

The fromBase64 internal function returns a value of type
std::optional&lt;std::pair&lt;size_t, Vector&lt;uint8_t&gt;&gt;&gt;. Returning std::nullopt indicates a decoding
failure. This patch changes it to return a tuple of std::tuple&lt;bool, size_t, Vector&lt;uint8_t&gt;&gt;,
where the initial bool determines whether to throw an error.

[1]: <a href="https://tc39.es/proposal-arraybuffer-base64/spec/#sec-uint8array.prototype.setfrombase64">https://tc39.es/proposal-arraybuffer-base64/spec/#sec-uint8array.prototype.setfrombase64</a>
[2]: <a href="https://tc39.es/proposal-arraybuffer-base64/spec/#sec-frombase64">https://tc39.es/proposal-arraybuffer-base64/spec/#sec-frombase64</a>
[3]: <a href="https://github.com/tc39/test262/blob/main/test/built-ins/Uint8Array/prototype/setFromBase64/writes-up-to-error.js">https://github.com/tc39/test262/blob/main/test/built-ins/Uint8Array/prototype/setFromBase64/writes-up-to-error.js</a>

* JSTests/stress/uint8array-setFromBase64-write-up-to-error.js: Added.
(shouldBe):
(shouldThrow):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.cpp:
(JSC::fromBase64):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/281174@main">https://commits.webkit.org/281174@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbb97e65cf17726e12a0ceec9ada8142a89a8715

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38351 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11516 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62693 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9466 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46002 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9670 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47764 "") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6756 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61053 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35884 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51066 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28622 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32600 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8343 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8470 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/52115 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54553 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8623 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64392 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58264 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2935 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8579 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55085 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2945 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51080 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55189 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2465 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80025 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8818 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34180 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13846 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35264 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36349 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35010 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me
Note that I'm now working on using imported SIMDUTD here, so soon the implementation gets large change.

sosukesuzuki committed:
[JSC] `Uint8Array#setFromBase64` should decode and write chunks which occur prior to bad data
https://bugs.webkit.org/show_bug.cgi?id=276859

Reviewed by Yusuke Suzuki.

According to the spec[1][2] and the test[3], Uint8Array.prototype.setFromBase64 throws a SyntaxError
when encountering bad data but writes the decoded chunks prior to bad data.

However, the current JSC discards the decoded data upon encountering bad data and throws an error.

The fromBase64 internal function returns a value of type
std::optional<std::pair<size_t, Vector<uint8_t>>>. Returning std::nullopt indicates a decoding
failure. This patch changes it to return a tuple of std::tuple<bool, size_t, Vector<uint8_t>>,
where the initial bool determines whether to throw an error.

[1]: https://tc39.es/proposal-arraybuffer-base64/spec/#sec-uint8array.prototype.setfrombase64
[2]: https://tc39.es/proposal-arraybuffer-base64/spec/#sec-frombase64
[3]: https://github.com/tc39/test262/blob/main/test/built-ins/Uint8Array/prototype/setFromBase64/writes-up-to-error.js

* JSTests/stress/uint8array-setFromBase64-write-up-to-error.js: Added.
(shouldBe):
(shouldThrow):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.cpp:
(JSC::fromBase64):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/281174@main
